parachain-upgrade/src/lib.rs

```diff
@@ -178,9 +166,21 @@ impl<T: Trait> ProvideInherent for Module<T> {
 	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
 
 	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
-		let vfp = data.get_data(&INHERENT_IDENTIFIER)
-			.and_then(|r| r.ok_or_else(|| "Validation Function Params inherent data not found".into()))
-			.expect("Gets and decodes vfp inherent data");
+		let vfp: ValidationFunctionParams = data.get_data(&INHERENT_IDENTIFIER)
+			.expect("Gets and decodes vfp inherent data")
+			.expect("Validation Function Params inherent data not found");
+
+
+		// initialization logic: we know that this runs exactly once every block,
+		// which means we can put the initialization logic here to remove the
+		// sequencing problem.
+		if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
```

Comment:
Can you elaborate a bit more on the context that you expect this initialization code to run in?

My understanding of the `create_inherent` function is that it's run only by the block author at the time of authoring the block. Furthermore, it is run in the context of a Runtime API instance, where this context is not the same as the one we create the block in the block.

As regular importing nodes of the chain will not run this logic, it seems the wrong place to put initialization behavior. Likewise, the changes made during the execution of this function on the authoring node will be immediately discarded and not used in the creation of the block.

Reply:
If that's the case, then this fix will not work. It was suggested [here](https://github.com/paritytech/substrate/issues/5757#issuecomment-618903497); I do not have enough knowledge of the system as is to say whether or not it does what it intends to in all cases. 

> the changes made during the execution of this function on the authoring node will be immediately discarded and not used in the creation of the block.

If that's the case, then an appropriate fix would be to move the pending validation function check into `Call::set_validation_function_parameters`; doing so should ensure those changes persist as part of the created inherent.

Reply:
> an appropriate fix would be to move the pending validation function check...

This turns out not to be true; when I make that change, tests fail. Returning the call apparently does not execute the call.

Reply:
A new attempt in e9cf4f8 appears to work to address this.